src/rpp/rpp/schedulers/trampoline_scheduler.hpp

```diff
@@ -42,52 +36,83 @@ struct scope_guard {
  */
 class trampoline final : public details::scheduler_tag
 {
+    class current_thread_schedulable : public details::schedulable
+    {
+    public:
+        current_thread_schedulable(time_point                  time_point,
+                                   std::invocable auto&&       fn,
+                                   rpp::composite_subscription subscription)
+            : schedulable(time_point, get_thread_local_id(), std::forward<decltype(fn)>(fn))
+            , m_subscription{std::move(subscription)} {}
+
+        bool is_subscribed() const { return m_subscription.is_subscribed(); }
+
+    private:
+        static size_t get_thread_local_id()
+        {
+            static thread_local size_t s_id;
+            return s_id++;
+        }
+
+    private:
+        rpp::composite_subscription m_subscription{};
+    };
+
+    /**
+     * \brief Returns optional thread_local schedulable queue. If optional has value -> someone just owns thread.
+     */
+    static std::optional<std::priority_queue<current_thread_schedulable>>& get_schedulable_queue()
+    {
+        static thread_local std::optional<std::priority_queue<current_thread_schedulable>> s_queue{};
+        return s_queue;
+    }
+
+    static void drain_queue()
+    {
+        auto& queue          = get_schedulable_queue();
+        auto  reset_at_final = utils::finally_action{[] { get_schedulable_queue().reset(); }};
+
+        while (!queue->empty())
+        {
+            const auto& top = queue->top();
+
+            auto function = wait_and_extract_executable_if_subscribed(top);
+
+            // firstly we need to pop schedulable from queue due to execution of function can add new schedulable
+            queue->pop();
+
+            if (function)
+                function();
+        }
+    }
+
+    [[nodiscard]] static std::function<void()> wait_and_extract_executable_if_subscribed(const  current_thread_schedulable& schedulable)
+    {
+        if (!schedulable.is_subscribed())
+            return {};
+
+        std::this_thread::sleep_until(schedulable.GetTimePoint());
```

Comment:
The `std::this_thread::sleep_until(schedulable.GetTimePoint());` is un-stoppable vs the stoppable `pop_with_wait(..., token)` in the previous implementation.
There could be this case the sharable subscription is held in other thread and somehow the subscription is unsubscribed in other thread. With the `::sleep_until()`, we can’t stop the thread responsively but have to wait for the thread to wake up and quit.

Can we use stoppable sleeping mechanism here?